emcc.py

```diff
@@ -1344,6 +1349,17 @@ def check(input_file):
             logger.warning('BINARYEN_ASYNC_COMPILATION requested, but disabled because of user options. ' + warning)
           elif 'BINARYEN_ASYNC_COMPILATION=0' not in settings_changes:
             logger.warning('BINARYEN_ASYNC_COMPILATION disabled due to user options. ' + warning)
+        # Swappable wasm module/asynchronous wasm compilation requires an indirect stub
+        # function generated to each function export from wasm module, so cannot use the
+        # concise form of grabbing exports that does not need to refer to each export individually.
+        if shared.Settings.SWAPPABLE_ASM_MODULE == 1 and not shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
+          shared.Settings.DECLARE_ASM_MODULE_EXPORTS = 1
+          logger.warning('Enabling -s DECLARE_ASM_MODULE_EXPORTS=1 since -s SWAPPABLE_ASM_MODULE=1 is used')
+        # Wasm -O3 builds use Meta-DCE which is currently not compatible with -s DECLARE_ASM_MODULE_EXPORTS=0 option.
+        if options.opt_level >= 3 or options.shrink_level >= 1 and not shared.Settings.DECLARE_ASM_MODULE_EXPORTS:
```

Comment:
instead of hardcoding assumptions about when metadce is run, how about erroring in the metadce code itself in tools/shared.py?

Reply:
I had added this check to both places: metadce has

```py
  def metadce(js_file, wasm_file, minify_whitespace, debug_info):
    logger.debug('running meta-DCE')
    assert Settings.DECLARE_ASM_MODULE_EXPORTS, 'Internal error: Meta-DCE is currently not compatible with -s DECLARE_ASM_MODULE_EXPORTS=0, build should have occurred with -s DECLARE_ASM_MODULE_EXPORTS=1 if we reach here!'
```

so this code in emcc.py frontend disables `DECLARE_ASM_MODULE_EXPORTS=0` if it's incompatible, but the metadce code then finally enforces that it doesn't get called with a currently incompatible flag.